Date: Nov 07 18:34:33 2017

From: b9ea321d93ec75ca98a2c7751590ca51bbdad036 Nov 07 18:36:11 2017

Subject: [PATCH] XRE-12038 - [AAMP] FF/REW support for DASH content

	Add support for retimestamping in qtdemux based on a
	custom event which is sent only when aamp is used.

Source: COMCAST


Upstream-Status: Pending

Signed-off-by: Christo Joseph <Christo_Joseph@comcast.com>

---

Index: gst-plugins-good-1.10.4/gst/isomp4/qtdemux.c
===================================================================
--- gst-plugins-good-1.10.4.orig/gst/isomp4/qtdemux.c
+++ gst-plugins-good-1.10.4/gst/isomp4/qtdemux.c
@@ -643,6 +643,11 @@ gst_qtdemux_init (GstQTDemux * qtdemux)
   qtdemux->flowcombiner = gst_flow_combiner_new ();
 
   GST_OBJECT_FLAG_SET (qtdemux, GST_ELEMENT_FLAG_INDEXABLE);
+
+  qtdemux->aamp_base_pts = GST_CLOCK_TIME_NONE;
+  qtdemux->aamp_base_pts_override = GST_CLOCK_TIME_NONE;
+  qtdemux->aamp_override_enabled = FALSE;
+  qtdemux->aamp_rate = 1.0;
 }
 
 static void
@@ -2187,7 +2192,7 @@ gst_qtdemux_handle_sink_event (GstPad *
       GST_DEBUG_OBJECT (demux, "Pushing newseg %" GST_SEGMENT_FORMAT, &segment);
 
       /* map segment to internal qt segments and push on each stream */
-      if (demux->n_streams) {
+      if (demux->n_streams && !demux->aamp_override_enabled) {
         if (demux->fragmented) {
           GstEvent *segment_event = gst_event_new_segment (&segment);
 
@@ -2199,7 +2204,9 @@ gst_qtdemux_handle_sink_event (GstPad *
           gst_qtdemux_map_and_push_segments (demux, &segment);
         }
       }
-
+      else if(demux->aamp_override_enabled) {
+         GST_WARNING_OBJECT (demux, "ignore newsegment %" GST_SEGMENT_FORMAT, &segment);
+      }
       /* clear leftover in current segment, if any */
       gst_adapter_clear (demux->adapter);
 
@@ -2296,6 +2303,37 @@ gst_qtdemux_handle_sink_event (GstPad *
       res = TRUE;
       goto drop;
     }
+    case GST_EVENT_CUSTOM_DOWNSTREAM:
+    {
+        const GstStructure *structure = gst_event_get_structure(event);
+        if (gst_structure_has_name(structure, "aamp_override"))
+        {
+            demux->aamp_base_pts = GST_CLOCK_TIME_NONE;
+            demux->aamp_rate = g_value_get_float(gst_structure_get_value(structure, "rate"));
+            demux->aamp_override_enabled = g_value_get_boolean(gst_structure_get_value(structure, "enable"));
+            g_print("%s:%d - aamp_override - enabled = %d, rate  %f \n", __FUNCTION__, __LINE__, demux->aamp_override_enabled, demux->aamp_rate);
+            if (demux->aamp_override_enabled)
+            {
+              const GValue * basePTSVal = gst_structure_get_value(structure, "basePTS");
+              if (basePTSVal)
+              {
+                demux->aamp_base_pts_override = g_value_get_uint64( basePTSVal);
+                g_print("%s:%d - aamp_base_pts_override %" G_GUINT64_FORMAT "\n", __FUNCTION__, __LINE__, demux->aamp_base_pts_override);
+              }
+              else
+              {
+                demux->aamp_base_pts_override = GST_CLOCK_TIME_NONE;
+              }
+            }
+            else
+            {
+              demux->aamp_base_pts_override = GST_CLOCK_TIME_NONE;
+            }
+            gst_event_unref (event);
+            goto drop;
+        }
+        break;
+    }
     default:
       break;
   }
@@ -5480,8 +5518,58 @@ gst_qtdemux_decorate_and_push_buffer (Gs
     goto exit;
   }
 
-  GST_BUFFER_DTS (buf) = dts;
-  GST_BUFFER_PTS (buf) = pts;
+  if (G_LIKELY (!qtdemux->aamp_override_enabled))
+  {
+      GST_BUFFER_DTS (buf) = dts;
+      GST_BUFFER_PTS (buf) = pts;
+  }
+  else
+  {
+      gdouble rate  = qtdemux->aamp_rate;
+      if(GST_CLOCK_TIME_NONE == qtdemux->aamp_base_pts)
+      {
+        GstSegment segment;
+        GST_BUFFER_DTS (buf) = dts;
+        GST_BUFFER_PTS (buf) = pts;
+        if (GST_CLOCK_TIME_NONE == qtdemux->aamp_base_pts_override)
+        {
+          qtdemux->aamp_base_pts = pts;
+        }
+        else
+        {
+          qtdemux->aamp_base_pts = qtdemux->aamp_base_pts_override;
+        }
+        g_print("%s:%d - aamp %f first pts %"G_GUINT64_FORMAT" ms\n",
+        __FUNCTION__, __LINE__, rate, GST_TIME_AS_MSECONDS(pts));
+        gst_segment_init(&segment, GST_FORMAT_TIME);
+        segment.start = 0;
+        GstEvent* event = gst_event_new_segment(&segment);
+        if (!gst_pad_push_event(stream->pad, event))
+        {
+          g_print("%s: gst_pad_push_event segment error\n", __FUNCTION__);
+        }
+      }
+      if ( rate > 0 )
+      {
+          if ((pts < qtdemux->aamp_base_pts) || (dts < qtdemux->aamp_base_pts))
+          {
+            gst_buffer_unref(buf);
+            goto exit;
+          }
+          GST_BUFFER_DTS (buf) = ( dts - qtdemux->aamp_base_pts)/rate;
+          GST_BUFFER_PTS (buf) = ( pts - qtdemux->aamp_base_pts)/rate;
+          GST_DEBUG_OBJECT (qtdemux, "aamp_trickmode %f orig pts %"G_GUINT64_FORMAT" restamped pts %"G_GUINT64_FORMAT" ms\n",
+                  rate, GST_TIME_AS_MSECONDS(pts), GST_TIME_AS_MSECONDS(GST_BUFFER_PTS (buf)));
+      }
+      else
+      {
+          rate = -rate;
+          GST_BUFFER_DTS (buf) = ( qtdemux->aamp_base_pts - dts)/rate;
+          GST_BUFFER_PTS (buf) = ( qtdemux->aamp_base_pts - pts)/rate;
+          GST_DEBUG_OBJECT (qtdemux, "aamp_trickmode %f orig pts %"G_GUINT64_FORMAT" restamped pts %"G_GUINT64_FORMAT" ms\n",
+                  -rate, GST_TIME_AS_MSECONDS(pts), GST_TIME_AS_MSECONDS(GST_BUFFER_PTS (buf)));
+      }
+  }
   GST_BUFFER_DURATION (buf) = duration;
   GST_BUFFER_OFFSET (buf) = -1;
   GST_BUFFER_OFFSET_END (buf) = -1;
Index: gst-plugins-good-1.10.4/gst/isomp4/qtdemux.h
===================================================================
--- gst-plugins-good-1.10.4.orig/gst/isomp4/qtdemux.h
+++ gst-plugins-good-1.10.4/gst/isomp4/qtdemux.h
@@ -232,6 +232,12 @@ struct _GstQTDemux {
    * header start.
    * Note : This is not computed from the GST_BUFFER_OFFSET field */
   guint64 fragment_start_offset;
+
+  /*Trick play support based on custom event*/
+  GstClockTime aamp_base_pts;
+  GstClockTime aamp_base_pts_override;
+  gdouble aamp_rate;
+  gboolean aamp_override_enabled;
 };
 
 struct _GstQTDemuxClass {
